### PR TITLE
feat(suite-native): Ripple send

### LINKED ISSUE #15369: Send XRP
Enable XRP send in Trezor Suite Lite app.

[Design](https://www.figma.com/design/sqKFZAIvk3zERxiWgthK8K/Send?node-id=2830-19660&node-type=canvas&t=vxkb3Ps0F8C1Ft2h-0)

### Acceptance criteria
- send flow works for XRP in the same was as for any other coin
- user is can input destination tag
- if the destination tag was submitted. there is an extra screen for reviewing the tag before the address review

### PR DESCRIPTION
**Pull request by PeKne**
Ripple-based coins send flow enabled.

## Description
- new destination tag input and its validation
- validation of XRP amount:
  - there has to stay at least 10 XRP after the transaction is send
  - the recipients account amount has to be equal or higher than 10XRP after the transaction
- there is an extra review screen if the user includes destination tag in the transaction
- a new `useHandleOnDeviceTransactionReview` created that abstracts the review process so it is possible to call it from both `SendDestinationTagReviewScreen` and `SendAddressReviewScreen` (depending if is the destination tag included)

## Related Issue

Resolve #15369 

## Screenshots:

https://github.com/user-attachments/assets/3429ac90-312a-4c4e-a3fb-d1b67dc2edc8

### CODE REVIEW

File: suite-native/module-send/src/components/AddressReviewStepList.tsx [hunk 1]
```diff
@@ -43,18 +41,17 @@ type NavigationProps = StackToStackCompositeNavigationProps<
...
     const [childHeights, setChildHeights] = useState<number[]>([]);
     const [stepIndex, setStepIndex] = useState(0);
-    const handleSendReviewFailure = useHandleSendReviewFailure({
-        accountKey,
-        transaction,
-        tokenContract,
-    });
+    const handleOnDeviceTransactionReview = useHandleOnDeviceTransactionReview();
     const setWasAppLeftDuringReview = useSetAtom(wasAppLeftDuringReviewAtom);
+    const isReviewInProgress = useSelector(
```
Comment (vytick) [inline suite-native/module-send/src/components/AddressReviewStepList.tsx]:
```suggestion
    const isOutputsReviewInProgress = useSelector(
```

File: suite-native/module-send/src/components/DestinationTagInput.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            style: {
+                alignItems: undefined,
+                flexDirection: 'column',
+                gap: utils.spacings.sp12,
+            },
+        },
+    }),
+);
+
+const INPUT_OPEN_HEIGHT = 88;
```
Comment (vytick) [inline suite-native/module-send/src/components/DestinationTagInput.tsx]: does this work correctly when system font size is bigger?

File: suite-native/module-send/src/components/DestinationTagInput.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+const INPUT_CLOSED_HEIGHT = 45;
+const ERROR_HEIGHT = 15;
```
Comment (vytick) [inline suite-native/module-send/src/components/DestinationTagInput.tsx]: does this work correctly when system font size is bigger?

File: suite-native/module-send/src/components/OutputsReviewFooter.tsx [hunk 4]
```diff
@@ -139,6 +139,7 @@ export const OutputsReviewFooter = ({
                         tokenSymbols: tokenSymbol ? [tokenSymbol] : undefined,
                         outputsCount: formValues.outputs.length,
                         selectedFee: formValues.selectedFee ?? 'normal',
+                        hasRippleDestinationTag: formValues.rippleDestinationTag ? true : false,
```
Comment (vytick) [inline suite-native/module-send/src/components/OutputsReviewFooter.tsx]:
```suggestion
                        hasRippleDestinationTag: !!formValues.rippleDestinationTag,
```

File: suite-native/module-send/src/hooks/useHandleOnDeviceTransactionReview.tsx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    const route = useRoute<RouteProps>();
+    const { accountKey, tokenContract, transaction } = route.params;
+    const navigation = useNavigation<NavigationProps>();
+    const { showAlert } = useAlert();
+    const isViewOnlyDevice = useSelector(selectIsDeviceRemembered);
+
+    const showReviewCancellationAlert = useShowReviewCancellationAlert();
+    const showDeviceDisconnectedAlert = useShowDeviceDisconnectedAlert();
+
+    const isReviewInProgress = useSelector(
```
Comment (vytick) [inline suite-native/module-send/src/hooks/useHandleOnDeviceTransactionReview.tsx]:
```suggestion
    const isOutputsReviewInProgress = useSelector(
```

File: suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+export const SendDestinationTagReviewScreen = ({
+    route,
+    navigation,
+}: StackProps<SendStackParamList, SendStackRoutes.SendDestinationTagReview>) => {
+    const { accountKey, tokenContract, destinationTag, transaction } = route.params;
+    const [wasReviewAlreadyStarted, setWasReviewAlreadyStarted] = useState(false);
+    const { translate } = useTranslate();
+
+    const isReviewInProgress = useSelector(
```
Comment (vytick) [inline suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx]: also here isOutputsReviewInProgress

File: suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import { SendConfirmOnDeviceImage } from '../components/SendConfirmOnDeviceImage';
+import { ReviewDestinationTagCard } from '../components/ReviewDestinationTagCard';
+import { useHandleOnDeviceTransactionReview } from '../hooks/useHandleOnDeviceTransactionReview';
+
+export const SendDestinationTagReviewScreen = ({
+    route,
+    navigation,
+}: StackProps<SendStackParamList, SendStackRoutes.SendDestinationTagReview>) => {
+    const { accountKey, tokenContract, destinationTag, transaction } = route.params;
+    const [wasReviewAlreadyStarted, setWasReviewAlreadyStarted] = useState(false);
```
Comment (vytick) [inline suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx]: nit^x: 
```suggestion
    const [hasReviewAlreadyStarted, setHasReviewAlreadyStarted] = useState(false);
```

File: suite-native/module-send/src/selectors.ts [hunk 8]
```diff
@@ -75,14 +75,39 @@ export const selectIsOutputsReviewInProgress = (
...
+    tokenContract?: TokenAddress,
+): boolean => {
+    const outputs = selectTransactionReviewOutputs(state, accountKey, tokenContract);
+    if (!outputs) return false;
+
+    return pipe(
+        outputs,
+        A.find(
+            output =>
+                // 'regular_legacy' is address of BTC accounts used in older firmware versions.
```
Comment (vytick) [inline suite-native/module-send/src/selectors.ts:105]: do we support those fw versions in mobile app?

File: suite-native/module-send/src/sendOutputsFormSchema.ts [hunk 9]
```diff
@@ -108,6 +122,47 @@ export const sendOutputsFormValidationSchema = yup.object({
                             return !isAmountDust(value, context);
                         },
                     )
+                    .test(
+                        'ripple-higher-than-reserve',
+                        'Amount is above the required unspendable reserve (10 XRP)',
```
Comment (vytick) [inline suite-native/module-send/src/sendOutputsFormSchema.ts:126]: Send max should probably work with limitation and not to prefill what cant be sent. Wdyt?

File: suite-native/module-send/src/components/DestinationTagInput.tsx (hunk 3 above)
Comment (vytick) [reply suite-native/module-send/src/components/DestinationTagInput.tsx]: It seems not to work correctly on narrow displays
Screenshot 2024-11-20 at 20 38 45

Comment (vytick) [reply suite-native/module-send/src/components/DestinationTagInput.tsx]: It probably shouldnt allow new line in the input
Screenshot 2024-11-20 at 20 41 27

File: suite-native/module-send/src/components/DestinationTagInput.tsx [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+const SCROLL_TO_DELAY = 200;
+
+export const DestinationTagInput = () => {
+    const inputWrapperRef = useRef<View | null>(null);
+    const inputRef = useRef<TextInput | null>(null);
+    const scrollView = useScrollView();
+    const { applyStyle } = useNativeStyles();
+    const [isInputDisplayed, setIsInputDisplayed] = useState(false);
```
Comment (vytick) [inline suite-native/module-send/src/components/DestinationTagInput.tsx]: It probably should be set to true when there is destination tag not null in the draft

File: suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    useEffect(() => {
+        if (isDestinationTagConfirmed) {
+            navigation.navigate(SendStackRoutes.SendAddressReview, {
+                accountKey,
+                tokenContract,
+                transaction,
+            });
+        }
+    }, [isDestinationTagConfirmed, accountKey, navigation, tokenContract, transaction]);
+
```
Comment (vytick) [inline suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx:62]: Tested on ios emulator and there is odd behaviour, that the tag can be reviewed on trezor device only if I tap on close the review sreen

Comment (vytick) [reply suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx:62]: https://github.com/user-attachments/assets/3907d4b7-7b84-4b40-b162-ac5b7f5206f0

File: suite-native/module-send/src/components/RecipientInputs.tsx [hunk 12]
```diff
@@ -1,21 +1,41 @@
...
+        selectAccountByKey(state, accountKey),
+    );
+
+    if (!account) return null;
+
+    const isRipple = account.networkType === 'ripple';
+
     return (
         <VStack spacing="sp16">
             <AddressInput index={index} accountKey={accountKey} />
```
Comment (vytick) [inline suite-native/module-send/src/components/RecipientInputs.tsx:28]: I have found odd behaviour when invalid address is saved in draft. When you make the address correct, validation still shows error:

https://github.com/user-attachments/assets/1b9c34b0-89fa-437a-a9df-d72c76de771a

**Review by vytick — CHANGES_REQUESTED** [review]

File: suite-native/module-send/src/components/AddressReviewStepList.tsx (hunk 1 above)
Comment (PeKne) [reply suite-native/module-send/src/components/AddressReviewStepList.tsx]: I did not like the `outputs` in the name. So I renamed the selector to `selectIsTransactionReviewInProgress` in 8b24ec3dbc and renamed all the selected values as well.

File: suite-native/module-send/src/components/OutputsReviewFooter.tsx (hunk 4 above)
Comment (PeKne) [reply suite-native/module-send/src/components/OutputsReviewFooter.tsx]: fixed 8b24ec3dbc

File: suite-native/module-send/src/hooks/useHandleOnDeviceTransactionReview.tsx (hunk 5 above)
Comment (PeKne) [reply suite-native/module-send/src/hooks/useHandleOnDeviceTransactionReview.tsx]: fixed 8b24ec3dbc

File: suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx (hunk 6 above)
Comment (PeKne) [reply suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx]: fixed 8b24ec3dbc

File: suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx (hunk 7 above)
Comment (PeKne) [reply suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx]: fixed 8b24ec3dbc

File: suite-native/module-send/src/selectors.ts (hunk 8 above)
Comment (PeKne) [reply suite-native/module-send/src/selectors.ts:105]: Yes, Trezor model One uses this old FW (version 1.x.x).

File: suite-native/module-send/src/components/DestinationTagInput.tsx (hunk 2 above)
Comment (PeKne) [reply suite-native/module-send/src/components/DestinationTagInput.tsx]: Static height removed in 3e7ad2680a.

File: suite-native/module-send/src/components/DestinationTagInput.tsx (hunk 3 above)
Comment (PeKne) [reply suite-native/module-send/src/components/DestinationTagInput.tsx]: Static height and all other mentioned problems fixed in 3e7ad2680a

File: suite-native/module-send/src/sendOutputsFormSchema.ts (hunk 9 above)
Comment (PeKne) [reply suite-native/module-send/src/sendOutputsFormSchema.ts:126]: send max fixed 3e7ad2680a

File: suite-native/module-send/src/components/DestinationTagInput.tsx (hunk 10 above)
Comment (PeKne) [reply suite-native/module-send/src/components/DestinationTagInput.tsx]: good point, fixed 3e7ad2680a

File: suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx (hunk 11 above)
Comment (PeKne) [reply suite-native/module-send/src/screens/SendDestinationTagReviewScreen.tsx:62]: There was this problem occurring after canceling the review for the first time. Then if you would try start review again it got stuck in this state. Fixed 3e7ad2680a

File: suite-native/module-send/src/components/RecipientInputs.tsx (hunk 12 above)
Comment (PeKne) [reply suite-native/module-send/src/components/RecipientInputs.tsx:28]: validation mode of prefilled form fixed in 3e7ad2680a

File: suite-native/module-send/src/sendOutputsFormSchema.ts [hunk 13]
```diff
@@ -133,13 +181,39 @@ export const sendOutputsFormValidationSchema = yup.object({
...
-                //       e.g. check if the amount is not higher than XRP reserve
             }),
         )
         .required(),
+    rippleDestinationTag: yup
+        .string()
+        .optional()
+        .matches(
+            /^\d*$/,
+            'You can only use positive numbers for the destination tag. fdsfds dsf dsfsf dsf dsf sdf dsf dsf .',
```
Comment (vytick) [inline suite-native/module-send/src/sendOutputsFormSchema.ts]:
```suggestion
            'You can only use positive numbers for the destination tag.',
```

File: suite-native/module-send/src/screens/SendAddressReviewScreen.tsx [hunk 14]
```diff
@@ -7,14 +7,13 @@ import { Translation, useTranslate } from '@suite-native/intl';
...
+    selectIsReceiveAddressOutputConfirmed,
+    selectIsTransactionReviewInProgress,
 } from '../selectors';
 import { AddressReviewStepList } from '../components/AddressReviewStepList';
 import { SendScreen } from '../components/SendScreen';
 import { SendScreenSubHeader } from '../components/SendScreenSubHeader';
 import { SendConfirmOnDeviceImage } from '../components/SendConfirmOnDeviceImage';
-import { useShowReviewCancellationAlert } from '../hooks/useShowReviewCancellationAlert';
 
 export const SendAddressReviewScreen = ({
```
Comment (vytick) [inline suite-native/module-send/src/screens/SendAddressReviewScreen.tsx:18]: I can get app to some weird state, when I randomly close review screens and try to do the send again


https://github.com/user-attachments/assets/07cb7c6d-2676-44ec-a76d-8a92ead969c5

**Review by vytick — CHANGES_REQUESTED** [review]
I have found two things, but otherwise it looks nice 👌

File: suite-native/module-send/src/sendOutputsFormSchema.ts (hunk 13 above)
Comment (PeKne) [reply suite-native/module-send/src/sendOutputsFormSchema.ts]: Damn, I was testing a long error message that can not fit in the screen width and forgot it there. Fixed 8839f5a337

File: suite-native/module-send/src/screens/SendAddressReviewScreen.tsx (hunk 14 above)
Comment (PeKne) [reply suite-native/module-send/src/screens/SendAddressReviewScreen.tsx:18]: There was `TrezorConnet.cancel('sign-tx')` called twice, which lead to this unconsistent behaviour. Fixed in 8839f5a337

**Review by vytick — APPROVED** [review]
Good job 👍  🚀

**Pull request merged**